###  pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

## Reading a data file

We can read pretty much all the data files like csv, xl, sql, HTML and many others and that too with a very similar format

In [2]:
## importing pandas

In [3]:
import pandas as pd

### Creating a data frame

In [17]:
dfx = pd.DataFrame({'passangerID':[1,2,3,4,5,6], 
       'Pclass':[3,3,3,2,2,1], 
       'Names': ['kelly coax','suresh sycophantic','ajanta raman','suresh raina','madhusudhan','rashmi talreja']
})
dfx

passangerID  Pclass               Names
0            1       3          kelly coax
1            2       3  suresh sycophantic
2            3       3        ajanta raman
3            4       2        suresh raina
4            5       2         madhusudhan
5            6       1      rashmi talreja

In [4]:
df = pd.read_csv('../Data_sets/Titanic/test.csv')

In [10]:
df.head(10) ## top 5 rows gets printed if nothing was specified

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   
5          897       3                    Svensson, Mr. Johan Cervin    male   
6          898       3                          Connolly, Miss. Kate  female   
7          899       2                  Caldwell, Mr. Albert Francis    male   
8          900       3     Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female   
9          901       3                       Davies, Mr. John Samuel    male   

    Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
0  34.5      0      0     330911   7.8292   NaN        Q  
1  47.0      1      0     363272   7.0000   NaN        S  
2  62.0      0      0     240276   9.6875   NaN        Q  
3  27.0      0      0     315154   8.6625   NaN        S  
4  22.0      1      1    3101298  12.2875   NaN        S  
5  14.0      0      0       7538   9.2250   NaN        S  
6  30.0      0      0     330972   7.6292   NaN        Q  
7  26.0      1      1     248738  29.0000   NaN        S  
8  18.0      0      0       2657   7.2292   NaN        C  
9  21.0      2      0  A/4 48871  24.1500   NaN        S

In [9]:
df.tail(10) ## bottom 5 rows are displayed if no number is specified

PassengerId  Pclass                                             Name  \
408         1300       3                  Riordan, Miss. Johanna Hannah""   
409         1301       3                        Peacock, Miss. Treasteall   
410         1302       3                           Naughton, Miss. Hannah   
411         1303       1  Minahan, Mrs. William Edward (Lillian E Thorpe)   
412         1304       3                   Henriksson, Miss. Jenny Lovisa   
413         1305       3                               Spector, Mr. Woolf   
414         1306       1                     Oliva y Ocana, Dona. Fermina   
415         1307       3                     Saether, Mr. Simon Sivertsen   
416         1308       3                              Ware, Mr. Frederick   
417         1309       3                         Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
408  female   NaN      0      0              334915    7.7208   NaN        Q  
409  female   3.0      1      1  SOTON/O.Q. 3101315   13.7750   NaN        S  
410  female   NaN      0      0              365237    7.7500   NaN        Q  
411  female  37.0      1      0               19928   90.0000   C78        Q  
412  female  28.0      0      0              347086    7.7750   NaN        S  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C

In [18]:
df.shape ## dimensions of data frame

(418, 11)

In [19]:
rows ,columns = df.shape
print("rows:", rows," columns:" ,columns)

rows: 418  columns: 11


#### which means 11 columns and 418 samples is what we are provided. Lets look at all the columns

In [7]:
df.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [11]:
df.describe

<bound method NDFrame.describe of      PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked 

In [13]:
df.dtypes ## Printing data types at one go

PassengerId      int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [14]:
df.describe() ## returns statastical analysis of data

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

## Using concepts of slicing and indexing

In [22]:
df[2:10]

PassengerId  Pclass                                          Name     Sex  \
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   
5          897       3                    Svensson, Mr. Johan Cervin    male   
6          898       3                          Connolly, Miss. Kate  female   
7          899       2                  Caldwell, Mr. Albert Francis    male   
8          900       3     Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female   
9          901       3                       Davies, Mr. John Samuel    male   

    Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
2  62.0      0      0     240276   9.6875   NaN        Q  
3  27.0      0      0     315154   8.6625   NaN        S  
4  22.0      1      1    3101298  12.2875   NaN        S  
5  14.0      0      0       7538   9.2250   NaN        S  
6  30.0      0      0     330972   7.6292   NaN        Q  
7  26.0      1      1     248738  29.0000   NaN        S  
8  18.0      0      0       2657   7.2292   NaN        C  
9  21.0      2      0  A/4 48871  24.1500   NaN        S

In [25]:
fare = df["Fare"]
print(fare.head())

0     7.8292
1     7.0000
2     9.6875
3     8.6625
4    12.2875
Name: Fare, dtype: float64


In [29]:
fare_ticket = df[["Fare","Ticket"]] ## Used to restrict data to a specific number of columns
print(fare_ticket.head())

      Fare   Ticket
0   7.8292   330911
1   7.0000   363272
2   9.6875   240276
3   8.6625   315154
4  12.2875  3101298


In [30]:
type(df['Ticket'])

pandas.core.series.Series

## Some Operations
- ### mean
- ### median
- ### max
- ### min
- ### std

In [34]:
print("Maximum age",df['Age'].max()," | Minimum age",df['Age'].min())

Maximum age 76.0  | Minimum age 0.17


## conditionaly selecting data from the data

In [38]:
x = df[df['Age'] >50]
print(x.shape)

(31, 11)


In [39]:
print(x.head())

    PassengerId  Pclass                                             Name  \
2           894       2                        Myles, Mr. Thomas Francis   
13          905       2                             Howard, Mr. Benjamin   
20          912       1                           Rothschild, Mr. Martin   
48          940       1  Bucknell, Mrs. William Robert (Emma Eliza Ward)   
69          961       1              Fortune, Mrs. Mark (Mary McDougald)   

       Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
2     male  62.0      0      0    240276    9.6875          NaN        Q  
13    male  63.0      1      0     24065   26.0000          NaN        S  
20    male  55.0      1      0  PC 17603   59.4000          NaN        C  
48  female  60.0      0      0     11813   76.2917          D15        C  
69  female  60.0      1      4     19950  263.0000  C23 C25 C27        S  


In [43]:
x = df[df.Ticket == df['Ticket'].max()]
x

PassengerId  Pclass                                               Name  \
14          906       1  Chaffee, Mrs. Herbert Fuller (Carrie Constance...   

       Sex   Age  SibSp  Parch       Ticket    Fare Cabin Embarked  
14  female  47.0      1      0  W.E.P. 5734  61.175   E31        S

In [48]:
## printing Pclass only 2 and 3
only_23 = df[df['Pclass'].isin([2,3])]
only_23.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [57]:
## another way
only_23 = df[(df["Pclass"] == 2) | (df["Pclass"] == 3)]
only_23.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [62]:
z = df[["PassengerId","Pclass","Sex","Embarked"]]
z[z['Sex'] == 'male']

PassengerId  Pclass   Sex Embarked
0            892       3  male        Q
2            894       2  male        Q
3            895       3  male        S
5            897       3  male        S
7            899       2  male        S
..           ...     ...   ...      ...
407         1299       1  male        C
413         1305       3  male        S
415         1307       3  male        S
416         1308       3  male        S
417         1309       3  male        C

[266 rows x 4 columns]

### this colud have been donne in a single line


In [63]:
z = df[["PassengerId","Pclass","Sex","Embarked"]][df['Sex'] == 'male']

In [64]:
z.head()

PassengerId  Pclass   Sex Embarked
0          892       3  male        Q
2          894       2  male        Q
3          895       3  male        S
5          897       3  male        S
7          899       2  male        S

### not null stuff

In [72]:
age_given = df[df['Age'].notna()]

In [74]:
age_given.shape

(332, 11)

In [75]:
df.shape

(418, 11)

## Lets get back to slicing operations

In [76]:
df[2:4]

PassengerId  Pclass                       Name   Sex   Age  SibSp  Parch  \
2          894       2  Myles, Mr. Thomas Francis  male  62.0      0      0   
3          895       3           Wirz, Mr. Albert  male  27.0      0      0   

   Ticket    Fare Cabin Embarked  
2  240276  9.6875   NaN        Q  
3  315154  8.6625   NaN        S

In [79]:
x = df.iloc[3:10 ,2:7]

In [80]:
x

Name     Sex   Age  SibSp  Parch
3                              Wirz, Mr. Albert    male  27.0      0      0
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1
5                    Svensson, Mr. Johan Cervin    male  14.0      0      0
6                          Connolly, Miss. Kate  female  30.0      0      0
7                  Caldwell, Mr. Albert Francis    male  26.0      1      1
8     Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0      0      0
9                       Davies, Mr. John Samuel    male  21.0      2      0

### here loc or iloc function allows us to select specific rows and columns. 
### difference b/w loc and iloc is that iloc works with indexes and thus is exclusive of the last value where as loc is inclusive

## Setting Indexes

In [86]:
df.index ## returns the current indexing scheme

RangeIndex(start=0, stop=418, step=1)

In [87]:
x = df.set_index('Name').shape

(418, 10)

In [88]:
x.head()

Name     Sex   Age  SibSp  Parch
3                              Wirz, Mr. Albert    male  27.0      0      0
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1
5                    Svensson, Mr. Johan Cervin    male  14.0      0      0
6                          Connolly, Miss. Kate  female  30.0      0      0
7                  Caldwell, Mr. Albert Francis    male  26.0      1      1

In [85]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

#### to change index in the main dataframe run inplace = 'True'